src/Adapter/SsiCache.php

```diff
@@ -42,56 +42,35 @@ class SsiCache implements CacheAdapterInterface
      */
     private $argumentResolver;
 
-    /**
-     * NEXT_MAJOR: make $argumentResolver mandatory when dropping sf < 3.1.
-     *
-     * @param string                           $token
-     * @param RouterInterface                  $router
-     * @param null|ControllerResolverInterface $resolver
-     * @param null|ArgumentsResolverInterface  $argumentResolver
-     */
-    public function __construct($token, RouterInterface $router, ControllerResolverInterface $resolver = null, ArgumentResolverInterface $argumentResolver = null)
-    {
-        if (interface_exists(ArgumentResolverInterface::class) && !$argumentResolver) {
-            @trigger_error(sprintf(
-                'Not providing a "%s" instance to "%s" is deprecated since 3.x and will not be possible in 4.0',
-                ArgumentResolverInterface::class,
-                __METHOD__
-            ), E_USER_DEPRECATED);
-        }
-
+    public function __construct(
+        string $token,
+        RouterInterface $router,
+        ?ControllerResolverInterface $resolver = null,
```

Comment:
3rd argument is null, what should I do with this? as this is major, can I replace the place with 4th?

Reply:
IMO if it is a parameter that can be not there, we should move it at the end, and not add the ? on the type hint.

? typehint means you can pass null
= null means you can not pass this parameter (but if you pass it, it should be instance of ControllerResolverInterface)

Reply:
You can also pass null as soon as you add `= null`, but if someone overrides such a method and changes the default value to something other than null, it fails.

https://3v4l.org/TuFIH

So IMO, if you add `= null`, you should also add `?`

Reply:
But it fails for another reason...

Saying a field is `= something` means that you can avoid passing this parameter and has a default value.

I see your point tho, on people misusing this = null, but not sure if we should support that.

Reply:
if you don't pass anything it is null, if you pass `null` it is still null, do not see anything bad about allowing `?`

Reply:
> But it fails for another reason... you are passing a null on a field you said it was an string
Yeah but it fails only for `b()`, not for `a()`, and the only thing that changes is the default value.

> Saying a field is = something means that you can avoid passing this parameter and has a default value.

Yeah but when something is null, the argument suddenly becomes nullable.

Reply:
Ok for the `?`, even though I find it meh, but it should be moved at the last argument, right?

Reply:
Wait… why don't we make both arguments mandatory and perserve the order exactly?

Reply:
https://github.com/sonata-project/SonataCacheBundle/blob/master/src/Adapter/SsiCache.php#L134 how is the 3rd even optional?!

Reply:
The plot thickens…

Reply:
Here is the introducing commit: 24fbb6818c0d6980c6d946f2f2456835f5c4b2f6 … no explanation, let's just make it mandatory too, shall we?

Reply:
done.